Restore ANSI win32 API functions

MaartenBent opened the pull request:
wxWidgets is always build in `UNICODE` using the `WIDE` win32 API functions. But applications using it might be built without `UNICODE`, using the `ANSI` win32 functions.

Right now wxWidgets `#undef`s some of the win32 API functions and defines them as inline functions using the `WIDE` win32 API. Making these unavailable for applications that want to use the `ANSI` versions later.

Fix this by inlining the functions as `WIDE` only when `UNICODE` was already defined, not when wxWidgets defines it themselves in `platform.h`.

See https://github.com/wxWidgets/wxWidgets/issues/25803

cc @jamarr81  would this work?

MaartenBent committed:
Remove unneeded winundef.h includes

And remove some probably unneeded undefs.

MaartenBent committed:
White-space and style cleanup in winundef.h

vadz commented on include/wx/msw/winundef.h:
```diff
@@ -15,30 +15,52 @@
+    #define wxLPSTR LPWSTR
+    #define wxLPWNDCLASS LPWNDCLASSW
+    #define wxDOCINFO DOCINFOW
+    #define wxLPFINDREPLACE LPFINDREPLACEW
```
Minor, but these could be `typedef`s (or even `using`) if we try to minimize the use of preprocessor (possibly not really helpful in this header...).

vadz reviewed (commented):
Thanks, I think this should work and I can't think of anything else that this would break, but I'd prefer to condense all the wrapper definitions using a helper macro before merging this.

MaartenBent committed:
Add option to restore ANSI win32 functions

Define wxNO_WIN32_W before including wxWidgets headers to inline
ANSI win32 API functions instead of WIDE win32 API functions.

Fixes #25803

MaartenBent committed:
Use WIDE variant of win32 API functions

wxWidgets uses the WIDE variant of the win32 API. But when the ANSI version of
Windows.h is included before wxWidgets headers and wxNO_WIN32_W is defined,
the ANSI variant of the functions will be used.

This can cause build errors in headers that are (indirectly) included by the user.
Because the class/string types do not match the function type.

So use the WIDE variant explicitly to prevent these errors.

jamarr81 commented:
@MaartenBent @vadz tysm for following up on this! I haven't had a chance to revisit it; however, these changes look good to me on paper.

vadz reviewed (approved):
Looks good, thanks!

It would be great if this could be tested in some application which needed this, but if nobody does it soon, I'll just merge this and hope that it works for them.

vadz commented:
I think it could be acceptable to do this in the existing `native.cpp` if we added a comment explaining that this is just for testing, but it would be nicer to do it in some other place. Maybe as part of the (GUI) test suite? We just need a test case calling some harmless Win32 function existing in both A and W versions, e.g. `GetWindowText()`.

MaartenBent committed:
CMake: Make _UNICODE/UNICODE defines private

Applications/libraries linking to wxWidgets libraries will not automatically inherit them anymore.

MaartenBent commented:
I added a separate test, building with and without `UNICODE`, that includes and uses `Windows.h`.

MaartenBent committed:
Add test for including Windows headers before wx headers

Test both ANSI and WIDE (with/without UNICODE) builds.
